resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCredentialRenewer.scala

```diff
@@ -0,0 +1,169 @@
+      }
+    }
+
+    val credentialRenewerRunnable =
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            val creds = getRenewedDelegationTokens(conf)
+            broadcastDelegationTokens(creds)
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
+              logWarning("Couldn't broadcast tokens, trying again in an hour", e)
+              credentialRenewerThread.schedule(this, 1, TimeUnit.HOURS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    scheduleRenewal(credentialRenewerRunnable)
+  }
+
+  private def getRenewedDelegationTokens(conf: SparkConf): Array[Byte] = {
+    logInfo(s"Attempting to login with ${conf.get(config.PRINCIPAL).orNull}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174
+    val ugi = if (mode == "keytab") {
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, secretFile)
+    } else {
+      UserGroupInformation.getUGIFromTicketCache(secretFile, principal)
+    }
+    val tempCreds = ugi.getCredentials
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    var nextRenewalTime = Long.MaxValue
+    ugi.doAs(new PrivilegedExceptionAction[Void] {
+      override def run(): Void = {
+        nextRenewalTime = tokenManager.obtainDelegationTokens(hadoopConf, tempCreds)
+        null
+      }
+    })
+
+    val currTime = System.currentTimeMillis()
+    timeOfNextRenewal = if (nextRenewalTime <= currTime) {
+      logWarning(s"Next credential renewal time ($nextRenewalTime) is earlier than " +
+        s"current time ($currTime), which is unexpected, please check your credential renewal " +
+        "related configurations in the target services.")
+      currTime
+    } else {
+      MesosCredentialRenewer.getNextRenewalTime(nextRenewalTime)
+    }
+    logInfo(s"Time of next renewal is $timeOfNextRenewal")
+
+    // Add the temp credentials back to the original ones.
+    UserGroupInformation.getCurrentUser.addCredentials(tempCreds)
+    SparkHadoopUtil.get.serialize(tempCreds)
+  }
+
+  private def broadcastDelegationTokens(tokens: Array[Byte]): Unit = {
+    // send token to existing executors
+    logInfo("Sending new tokens to all executors")
+    driverEndpoint.send(UpdateDelegationTokens(tokens))
+  }
+}
+
+object MesosCredentialRenewer extends Logging {
+  def getTokenRenewalTime(bytes: Array[Byte], conf: SparkConf): Long = {
+    val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+    val creds = SparkHadoopUtil.get.deserialize(bytes)
+    val renewalTimes = creds.getAllTokens.asScala.flatMap { t =>
+      Try {
+        t.renew(hadoopConf)
+      }.toOption
+    }
+    if (renewalTimes.isEmpty) Long.MaxValue else renewalTimes.min
+  }
+
+  def getNextRenewalTime(t: Long): Long = {
```

Comment:
This looks similar but not exactly the same as the logic in YARN's `writeNewCredentialsToHDFS`. Both should be using the same logic to calculate these things, so probably time for some refactoring.